### ISSUE TITLE
VFS2: `git clone` fails with internal error

### ISSUE THREAD

**Issue by fvoznika** [google/gvisor#3345]
Repro:
```
docker run --rm --runtime=runsc-vfs2 --entrypoint /bin/bash -ti l.gcr.io/google/bazel -c 'git clone https://github.com/google/gvisor.git'
```

Error:
```
error: internal error: refs/remotes/origin/dependabot/bundler/benchmarks/workloads/ruby/puma-3.12.6 is not a valid packed reference!
error: internal error: refs/remotes/origin/go is not a valid packed reference!
error: internal error: refs/remotes/origin/master is not a valid packed reference!
error: internal error: refs/remotes/origin/test/cl253260121 is not a valid packed reference!
error: internal error: refs/remotes/origin/test/cl258263294 is not a valid packed reference!
fatal: update_ref failed for ref 'HEAD': cannot update the ref 'HEAD': Trying to write ref refs/heads/master with nonexistent object 6f7f73996791bbab6b63c248000df0d3ce652f2b

```

**Comment by ayushr2** [comment]
So a part of the issue seems to be that `link(2)` has not been implemented for vfs2.
Fixing that does not solve this issue. I _think_ the real issue is that some additions to the dentry tree are not being echoed by `readdir(2)`.
Here is a simple example of how you can repro a simpler issue:
VFS2
```
$ docker run --runtime=vfs2-vfs2-d -it ubuntu bash
root@cd3257bf14a4:/# mkdir temp
root@cd3257bf14a4:/# cd temp/
root@cd3257bf14a4:/temp# ls
root@cd3257bf14a4:/temp# touch test
root@cd3257bf14a4:/temp# ls
```

VFS1
```
$ docker run --runtime=vfs2-d -it ubuntu bash
root@209bde31dbdf:/# mkdir temp
root@209bde31dbdf:/# cd temp/
root@209bde31dbdf:/temp# ls
root@209bde31dbdf:/temp# touch test
root@209bde31dbdf:/temp# ls
test
```

**Comment by ayushr2** [comment]
Seems to be an issue with the fact that `d.dirents` is never updated when children are added.

**Comment by ayushr2** [comment]
Should be fixed by #3395
